### ISSUE TITLE
[Feature request] Single-line macro syntax

### ISSUE THREAD

**Issue by Rangi42** [gbdev/rgbds#902]
Macros need at least three lines, since the `MACRO` and `ENDM` have to be on their own. This is kind of verbose for small macros, and the documentation even suggests using `EQUS` instead for brevity. However, that's "cheating" since it's not really a macro invocation; it can't take arguments, and relies on `EQUS` expansion (arguably a misfeature).

I propose a single-line `DEF mac MACRO ...` syntax for cases like this. After eating the `MACRO` token, the parser would call a lexer function to capture everything up to the next newline/end-of-file as a string. This would only allow single-statement macros, but would combine neatly with #901 for multiple statements.

Some examples:

```
DEF lb MACRO ld \1, (\2) << 8 | (\3)

; instead of pokecrystal's `hlcoord EQUS "coord hl,"`
DEF hlcoord MACRO coord hl, \#

; from rgbds-structs
DEF bytes MACRO new_field \1, RB, \2

; the "small one-line macro" example from the docs, with '.' as a statement separator
DEF pusha MACRO push af . push bc . push de . push hl

; or '\' as a statement separator
DEF wait_vblank MACRO : \ ldh a, [rSTAT] \ and STATF_BUSY \ jr nz, :-
```

Basically these would be equivalent:

```
DEF mac MACRO ...

MACRO mac
    ...
ENDM
```

**Comment by aaaaaa123456789** [comment]
Throwing this out there early: how would this interact with `REDEF` and traditional macro definitions?

**Comment by Rangi42** [comment]
You can `PURGE` and redefine macros, so `REDEF mac MACRO ...` should work like that.

**Comment by ISSOtm** [comment]
I don't think the added complexity is worth saving two lines per such macro.

**Comment by aaaaaa123456789** [comment]
This feature came up in a debate about replacing current uses of `EQUS`, so keep that in mind.

**Comment by Rangi42** [comment]
It wouldn't really be complex at all to implement. The parser rule is trivial:

```
def_macro	: def_id T_POP_MACRO {
			$<captureTerminated>$ = lexer_CaptureDefMacroLine(&captureBody);
		} endofline {
			if ($<captureTerminated>6)
				sym_AddMacro($1, captureBody.lineNo, captureBody.body,
					     captureBody.size);
		}
;
```

And the lexer function is mostly copied from `lexer_CaptureRept` and `lexer_CaptureMacroBody`. (The boilerplate of initializing and finalizing the capture body might be better off factored out into utility functions, since this would be the third repetition of it.)

```
bool lexer_CaptureDefMacroLine(struct CaptureBody *capture)
{
	capture->lineNo = lexer_GetLineNo();
	capture->body = startCapture();

	int c = EOF;

	/* If the file is `mmap`ed, we need not to unmap it to keep access to the macro */
	if (lexerState->isMmapped)
		lexerState->isReferenced = true;

	/* Capture up to a newline or EOF */
	do {
		c = nextChar();
	} while (c != '\n' && c != '\r' && c != EOF);

	if (c == '\n' || c == '\r') {
		if (c == '\r' && peek() == '\n') {
			shiftChar();
			lexerState->captureSize--;
		}
		lexerState->captureSize--;
	}

	// This being NULL means we're capturing from the capture buf, which is `realloc`'d during
	// the whole capture process, and so MUST be retrieved at the end
	if (!capture->body)
		capture->body = lexerState->captureBuf;
	capture->size = lexerState->captureSize;
	lexerState->capturing = false;
	lexerState->captureBuf = NULL;
	lexerState->disableMacroArgs = false;
	lexerState->disableInterpolation = false;
	lexerState->atLineStart = false;

	/* Returns true if a newline terminated the block, false if it reached EOF first */
	return c != EOF;
}
```

(Haven't tested that, but I think it would work.)

**Comment by Rangi42** [comment]
Actually, even without this feature I think we should factor out an `endCapture` function to pair with the existing `startCapture`

**Comment by ISSOtm** [comment]
That's still some additional maintaining complexity due to more code, which I believe is not worth the associated feature. `EQUS` is not meant to be an alternative to macros, it's only a more efficient shorthand whenever things line up right. (Hell, I'm pretty sure you could leverage old-style macro definition and `ONE_LINER equs "MACRO\n"` to squeeze out the first line.)

Proper "in-line" macros are separate feature, which have been discussed, and I don't think `EQUS` is close to a good fit for them.

**Comment by Rangi42** [comment]
I agree that `EQUS` is a poor substitute for short macros, but rgbasm(5) specifically recommends it for "small one-line macros", and some projects do use it that way (like `text EQUS "db 0,"`). At the very least the documentation should be updated to advise against this: its own example, `DEF pusha EQUS "push af\npush bc\npush de\npush hl\n"`, ought to be a macro with a four-line body. Maybe it could recommend an inline usage like `DEF tiles EQUS "* 16"` (as used in both pret and gb-open-world) until user-defined functions can do that better.

**Comment by ISSOtm** [comment]
You're right that the documentation *definitely* needs updating. Using multiple lines in an `EQUS` is largely an "advanced" feature, which interacts with the lexer and parser in non-obvious ways, so the documentation especially should not promote it. (Though very few people seem to read the documentation, lol.)

**Comment by aaaaaa123456789** [comment]
> That's still some additional maintaining complexity due to more code, which I believe is not worth the associated feature. `EQUS` is not meant to be an alternative to macros, it's only a more efficient shorthand whenever things line up right. (Hell, I'm pretty sure you could leverage old-style macro definition and `ONE_LINER equs "MACRO\n"` to squeeze out the first line.)
> 
> Proper "in-line" macros are separate feature, which have been discussed, and I don't think `EQUS` is close to a good fit for them.

Whether a feature was meant to be used for a purpose and whether it _is_ used for that purpose are completely separate matters. When you pull the rug under projects by removing a feature because you don't like what it is being used for and don't add a replacement, you create a problem for everyone. (This is why I mentioned this feature was brought up in the context of #905; it isn't really that interesting unless that is being planned.)

[This file](https://github.com/pret/pokecrystal/blob/master/macros/legacy.asm) is a perfectly good example of why single-line macros would be desirable; you can easily imagine that file becoming completely unmanageable if each `EQUS` was replaced by a three-line macro expansion. #901 might have helped in that regard, but I can't even imagine how that would interact with macro definitions in any sane way at all, and I'm pretty sure the only sane way to handle regular macros would be to require that `ENDM` appears on a line by itself. 

All that being said, this is pretty much a non-issue unless #905 is being considered. As long as that's not on the radar, `EQUS` covers this use case pretty well.

**Comment by ISSOtm** [comment]
We will get to alternatives to `EQUS` when we get there. Since nobody so far has had the meantime to draft a replacement to `EQUS`, it's here to stay.

**We will not remove a feature until an acceptable alternative exists for its use cases.**
I don't know how many times I will have to emphasize this so people will stop complaining about us "pulling the rug"&mdash;we aren't, and we won't.

**Comment by Rangi42** [comment]
Oh hey, Ruby has something like this: https://zverok.space/blog/2023-12-01-syntax-sugar5-endless-methods.html

**Comment by Rangi42** [comment]
With unlimited-length strings, we could do this:

```asm
MACRO DEF_MACRO
	REDEF mac_name EQUS "\1"
	shift
	REDEF mac_body EQUS "\#"
	REDEF mac_inner EQUS "MACRO {mac_name}\n\t{mac_body}\nENDM"
	{mac_inner}
ENDM

	DEF_MACRO text, "db \"<TEXT>\", "
	DEF_MACRO pushall, push af :: push bc :: push de :: push hl
	DEF_MACRO faceperson, faceobject

**Comment by Rangi42** [comment]
Until/unless we remove `EQUS` expansion (#905), there's no real need for single-line macro syntax; and by the time we do that, we'll probably have better ideas for how to do this. (By then we may also have named macro args, or other additions that wouldn't work well with this `DEF name MACRO body` proposal.) So I'm closing this.
